sw/host/ot_certs/src/offsets.rs

```diff
@@ -0,0 +1,233 @@
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Context, Result};
+use memchr::memmem;
+use serde::Serialize;
+use std::collections::HashMap;
+
+use crate::template::{Conversion, Value, Variable, VariableType};
+use crate::types::{convert_type, Type};
+
+#[derive(Debug, Clone, Serialize)]
+pub struct CertificateOffset {
+    /// Offset into the DER.
+    pub offset: usize,
+    /// Type of the variable in the DER.
+    pub target_type: VariableType,
+    /// Conversion that needs to be applied to the original variable.
+    pub conversion: Option<Conversion>,
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct CertificateVariable {
+    pub name: String,
+    pub source_type: VariableType,
+    pub offsets: Vec<CertificateOffset>,
+}
+
+/// X509 certificate in DER form where the offset to variables has been recorded.
+#[derive(Debug, Clone, Serialize)]
+pub struct CertificateWithVariables {
+    /// Name of the certificate, will be used to generate names in the template.
+    pub name: String,
+    /// Certificate in binary form.
+    pub cert: Vec<u8>,
+    /// List of offsets where the variables are used. A variable can appear multiple
+    /// times with different types.
+    pub variables: Vec<CertificateVariable>,
+}
+
+struct GeneratorVariable {
+    hidden: bool,
+    source_type: VariableType,
+    values: Vec<Vec<u8>>,
+    offsets: Vec<CertificateOffset>,
+}
+
+impl GeneratorVariable {
+    fn new(source_type: VariableType) -> GeneratorVariable {
+        GeneratorVariable {
+            hidden: false,
+            source_type,
+            values: Vec::new(),
+            offsets: Vec::new(),
+        }
+    }
+}
+
+// The generators holds the information about the offset to the variables
+// in the certificate and the randomly generated values that we used during generation.
+pub struct OffsetGenerator {
+    variables: HashMap<String, GeneratorVariable>,
+}
+
+impl OffsetGenerator {
+    pub fn new(variables: &HashMap<String, VariableType>) -> OffsetGenerator {
+        let variables = variables
+            .iter()
+            .map(|(name, source_type)| (name.clone(), GeneratorVariable::new(*source_type)))
+            .collect();
+        OffsetGenerator { variables }
+    }
+
+    fn add_variable(&mut self, bytes: Vec<u8>, target_type: VariableType, val: &Variable) {
+        let Variable { name, convert } = val;
+        // Register value.
+        if let Some(entry) = self.variables.get_mut(name) {
+            entry.values.push(bytes);
+            // Register offset.
+            entry.offsets.push(CertificateOffset {
+                offset: 0,
+                target_type,
+                conversion: *convert,
+            });
+        } else {
+            // This should never occur because the variable name should have been checked before
+            // in get_value.
+            panic!("variable '{}' does not exist", name);
```

Comment:
When called, the variable is known to exist because one has to lookup information before, hence the panic. Also `add_variable` is not public. I would prefer to keep it this way.